Apache DataSketches plugin for Trino

ShashwatArghode opened the pull request:
Plugin to query Apache Datasketches (https://datasketches.apache.org/docs/Theta/ThetaSketchFramework.html). This includes merge and estimate function for theta sketch.

Praveen2112 commented on plugin/trino-datasketches/src/main/java/io/trino/plugin/datasketches/state/SketchStateFactory.java:
```diff
@@ -0,0 +1,221 @@
+        return GroupedSketchState.class;
+    }
+
+    public static class GroupedSketchState
```
Can we move `SingleSketchState` and `GroupedSketchState` to a new file ?

Praveen2112 replied:
Can we move them to a new file ?

Praveen2112 commented on plugin/trino-datasketches/src/main/java/io/trino/plugin/datasketches/state/SketchStateSerializer.java:
```diff
@@ -0,0 +1,70 @@
+        else {
+            Slice slice = state.getSketch();
+
+            SliceOutput sliceOutput = Slices.allocate(SIZE_OF_LONG + SIZE_OF_INT + SIZE_OF_INT + slice.length()).getOutput();
```
```suggestion
            SliceOutput sliceOutput = new DynamicSliceOutput(SIZE_OF_LONG + SIZE_OF_INT + SIZE_OF_INT + slice.length());
```

ShashwatArghode replied:
Done.

Praveen2112 commented on plugin/trino-datasketches/src/main/java/io/trino/plugin/datasketches/theta/Estimate.java:
```diff
@@ -0,0 +1,64 @@
+
+        Sketch sketch = Sketches.wrapSketch(Memory.wrap(inputValue.getBytes()), seed);
+
+        return sketch.getEstimate();
```
Can we inline sketch ?

ShashwatArghode replied:
Done.

Praveen2112 reviewed (commented):
Thanks for working on this. Can we have some tests for like validating the result when we hit it to Trino. And a docs describing the functions would also be helpful.

Praveen2112 commented on plugin/trino-datasketches/src/main/java/io/trino/plugin/datasketches/state/SketchStateFactory.java:
```diff
@@ -0,0 +1,221 @@
+        return GroupedSketchState.class;
+    }
+
+    public static class GroupedSketchState
```
Since they are more details to capture, can we move it to a new file ? WDYT ?

ShashwatArghode replied:
I think keeping the structure consistent with other state factories would make more sense. Ex. HyperLogLog, LongApproximateMostFrequent, NumericHistogram etc.

Praveen2112 commented on plugin/trino-datasketches/src/main/java/io/trino/plugin/datasketches/theta/Union.java:
```diff
@@ -0,0 +1,70 @@
+     * @param state Current state of the sketch
+     * @param inputValue Input to add to the aggregate
+     */
+
```
We can remove this line

ShashwatArghode replied:
Done.

Praveen2112 commented:
Please change the commit message to Trino

ShashwatArghode commented:
Apologies for the delay in getting back to this. I have addressed all the comments and added the doc as well. Thanks for reviewing the PR, do let me know if it needs more changes.

ShashwatArghode commented on docs/src/main/sphinx/connector.rst:
```diff
@@ -13,6 +13,7 @@ from different data sources.
     Black Hole      <connector/blackhole>
     Cassandra       <connector/cassandra>
     ClickHouse      <connector/clickhouse>
+    DataSketches    <connector/datasketches>
```
Fixed it.

ShashwatArghode commented on plugin/trino-datasketches/src/test/java/io/trino/plugin/datasketches/theta/MergeEstimateTest.java:
```diff
@@ -0,0 +1,82 @@
+        union.update(sketch2);
+        Sketch unionResult = union.getResult();
+
+        assertEquals(unionResult.getEstimate(), estimate, 0);
```
Looked into TestMLQueries and adding test cases using Trino queries, but, for that, we will either need to support the creation of datasketches in Presto or register Hive UDFs for test cases for inserting Sketch in the database in binary format. We can add this when we add support for the creation of Sketches in Presto.

ebyhr replied:
As we talked in Slack, it should be testable without such supports. We can prepare Sketches data in Java library and then insert to varbinary column at this time.

ShashwatArghode replied:
`BytesWritable` from hadoop turned out to be the key to insert the sketches in binary format in the database. Added the test case.

ShashwatArghode commented:
@ebyhr, Do let me know if any more changes are needed to this.

bitsondatadev commented:
:wave: @ShashwatArghode - this PR has become inactive. If you're still interested in working on it, please let us know, and we can try to get reviewers to help with that.

We're working on closing out old and inactive PRs, so if you're too busy or this has too many merge conflicts to be worth picking back up, we'll be making another pass to close it out in a few weeks.

ShashwatArghode commented:
Hi @bitsondatadev,
All the old review comments have been addressed, we can try to get reviewers to move this forward. I will resolve the conflicts in the meantime.

findepi commented on plugin/trino-datasketches/src/main/java/io/trino/plugin/datasketches/state/SketchState.java:
```diff
@@ -0,0 +1,40 @@
+public interface SketchState
+        extends AccumulatorState
+{
+    Slice getSketch();
```
https://github.com/trinodb/trino/pull/14290 added a Theta sketch aggregation function: `IcebergThetaSketchForStats`
I forgot about this PR (and it is still waiting for "syntax review"), but anyway that function is for Iceberg plugin's internal use. Also this happens to be sketch building aggregation function that's not being added here.

Anyway, i was able to  use `UpdateSketch`  directly in the state class.
For sketch union aggregation, the state could probably hold directly onto an `org.apache.datasketches.theta.Union` object

findepi commented on plugin/trino-datasketches/src/main/java/io/trino/plugin/datasketches/theta/UnionWithParams.java:
```diff
@@ -0,0 +1,69 @@
+    private UnionWithParams() {}
+
+    @InputFunction
+    public static void input(@AggregationState SketchState state, @SqlType(StandardTypes.VARBINARY) Slice inputValue, @SqlType(StandardTypes.INTEGER) Integer normEntries, @SqlType(StandardTypes.BIGINT) Long seed)
```
What does normEntries mean?

findepi commented on plugin/trino-datasketches/src/main/java/io/trino/plugin/datasketches/theta/UnionWithParams.java:
```diff
@@ -0,0 +1,69 @@
+    {
+        state.setNominalEntries(normEntries);
+        state.setSeed(seed);
+        state.setSketch(inputValue);
```
Can this overwrite the state/result after previous `input` function invocation?

findepi commented on docs/src/main/sphinx/functions/datasketches.rst:
```diff
@@ -0,0 +1,39 @@
+DataSketches functions
+----------------------
+
+.. function:: theta_sketch_union(sketches) -> sketch
```
there are two variants of `theta_sketch_union`, one having parameters. The parameters need to be documented.

findepi replied:
also, should the docs say this is an aggregation function?

findepi replied:
> `-> sketch`

there is no "sketch" type. should this be "varbinary"?

findepi commented:
I was just skimming.  Not reviewing until the syntax is reviewed (https://github.com/trinodb/trino/pull/6643#event-4222400814)

vgankidi commented:
Hi, could I ask what open concerns are there on the syntax? I couldn't find relevant context from the comment above.

bitsondatadev commented:
Hey @vgankidi, syntax needs review needs a particular review generally from @martint, @electrum, or @dain. 

This just means that there needs to be a review around the consensus on how things are being done. I believe this is something we should work on defining or making a little more transparent.

mosabua commented:
:wave: @ShashwatArghode  - this PR is inactive and doesn't seem to be under development. If you'd like to continue work on this at any point in the future, feel free to re-open.
